feat: Ship builders / ship factories

petervdmeer opened the pull request:
**Feature:** This PR implements the ships builder as discussed in #6204

## Feature Details
This adds a ships-builder class as discussed in comment https://github.com/endless-sky/endless-sky/pull/6204#pullrequestreview-735453664. The ships-builder class is called ShipsFactory.cpp for now, since I was thinking of the factory design pattern when implementing this, but I'm okay with renaming it to ShipBuilder.cpp if that is desired.
The ships-builder class handles loading of Ships; moving the loading-functions from ships to the ships-builder class allows ship to get rid of the dependency on GameData.

## UI Screenshots
N/A (all of the changes here are under the hood)

## Usage Examples
N/A

## Testing Done
I loaded a savegame, checked that all ships were loaded properly. And checked the ships in stock, they were also all present and showed the proper attributes.

## Performance Impact
No performance impact is expected for the majority of this change.

quyykk reviewed (commented):
Instead of having a million setters in Ship I think it's cleaner if ShipFactory is a friend of Ship.

petervdmeer commented:
> Instead of having a million setters in Ship I think it's cleaner if ShipFactory is a friend of Ship.

Yes, I added the friend-class and removed the new Set-functions, except for the ExplosionEffects, because they do a bit more than just setting data.

tehhowch reviewed (changes requested):
Thanks for taking a stab at this! The prompting comment of mine wasn't very detailed; the gist is we want to distinguish between "templates" (DataNode collections / "plain data") and "runtime models".

petervdmeer commented on source/ShipsFactory.cpp:
```diff
@@ -0,0 +1,377 @@
+
+// Load a ship from a datafile. Creation of the Ship object itself has
+// already been done by the caller. Getting only the DataNode as parameter
+// and returning a shared_ptr<Ship> is nicer than getting the ship as
+// parameter by reference, but the ES code allows overwriting ship-
+// definitions by a new load, so we need to support overwriting existing
+// ship definitions here.
```
That is quite nice indeed.

petervdmeer replied:
I thought about this a bit more (and tried to write some code for this); if we split Ships into a "ShipTemplate" (new class) and a "ShipRuntime" (the current Ship class), then we probably also need to split other classes like Mission into a "MissionTemplate" (a template that refers to ShipTemplates for to-be-instantiated NPCs) and a "MissionInstantiated" (an active mission that the player is actively doing, with NPCs being actual ships). Do you agree?

This PR here was originally created to separate the DataFile/SaveGame-loading from the runtime (and future template classes), so removing the GameData and UniverseObjects dependencies from the RunTime Ship.
I feel both the split of Ships into ShipRuntimes and  ShipTemplates and the removal of the GameData and UniverseObject dependencies from Runtimes and Templates (by making the load functions external) are valuable. What do you think?

tehhowch replied:
All `Load` members would eventually be split.

I want us to have a separate library that handles the loading of text files into templates so that we can share the game's parser and serializer implementations with the community.

So this library would convert DataNodes into some class-specific representation of the data, and handle writing that back into DataNodes.

petervdmeer commented on source/UniverseObjects.cpp:
```diff
@@ -122,8 +123,9 @@ void UniverseObjects::FinishLoading()
 	UpdateSystems();
 
 	// And, update the ships with the outfits we've now finished loading.
+	const auto &sf = ShipsFactory(*this);
```
Providing the UniverseObjects context (the `this` in the code here) will be needed somewhere in the process of translating from a "template" to an "instantiated runtime model" to add pointers to "related universe runtime data" like outfits, systems, planets, graphics, effects.

petervdmeer commented on source/ShipsFactory.cpp:
```diff
@@ -0,0 +1,385 @@
+		ship.base = GameData::Ships().Get(ship.ModelName());
+		ship.variantName = node.Token(2);
+	}
+	ship.isDefined = true;
```
Right, we might even want to delay the creation of the "runtime" data models until all other loading is done, so that we don't even need a FinishLoading() anymore.

petervdmeer commented on source/Ship.h:
```diff
@@ -553,21 +580,19 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
+	const Effect *effectLeakageSpark = nullptr;
+	const Effect *effectBurningSpark = nullptr;
+	const Effect *effectSmoke = nullptr;
+	const Effect *effectJumpDrive = nullptr;
```
I used the struct option. Maybe the struct will move to the ShipModel in a future update, but if that happens, then that will be part of a larger change, not specific to this comment.

tibetiroka commented on source/ShipModel.cpp:
```diff
@@ -0,0 +1,25 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

tibetiroka commented on source/ShipModelLoader.cpp:
```diff
@@ -0,0 +1,59 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

tibetiroka commented on source/ShipModelLoader.h:
```diff
@@ -0,0 +1,51 @@
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

Amazinite commented:
@petervdmeer What's the status of this refactor?

petervdmeer commented:
> @petervdmeer What's the status of this refactor?

I still feel that this is a valuable improvement to the codebase, but I don't have time to work on this at the moment. I'll close this PR for now, and re-open it once I have some time available again.
